anvil/src/shell.rs

```diff
@@ -45,6 +46,56 @@ pub type MyWindowMap = WindowMap<
 
 pub type MyCompositorToken = CompositorToken<Roles>;
 
+struct MoveSurfaceGrab {
+    window_map: Rc<RefCell<MyWindowMap>>,
+    toplevel: SurfaceKind<Roles>,
+    initial_pointer_location: (f64, f64),
+    initial_window_location: (i32, i32),
+}
+
+impl PointerGrab for MoveSurfaceGrab {
+    fn motion(
+        &mut self,
+        _handle: &mut PointerInnerHandle<'_>,
+        location: (f64, f64),
+        _focus: Option<(wl_surface::WlSurface, (f64, f64))>,
+        _serial: u32,
+        _time: u32,
+    ) {
+        let dx = location.0 - self.initial_pointer_location.0;
+        let dy = location.1 - self.initial_pointer_location.1;
+        let new_window_x = (self.initial_window_location.0 as f64 + dx) as i32;
+        let new_window_y = (self.initial_window_location.1 as f64 + dy) as i32;
+
+        self.window_map
+            .borrow_mut()
+            .set_location(&self.toplevel, (new_window_x, new_window_y));
+    }
+
+    fn button(
+        &mut self,
+        handle: &mut PointerInnerHandle<'_>,
+        button: u32,
+        state: ButtonState,
+        serial: u32,
+        time: u32,
+    ) {
+        // TODO: should use the button that initiated the grab, not hardcode left.
```

Comment:
Indeed, and this will require some changes in Smithay to expose this from the implicit grab that is generated whenever a button is pressed, so that that implicit `ClickGrab` can be upgraded into a `MoveSurfaceGrab`.